### PR TITLE
Csum neumaier

### PR DESCRIPTION
**Pull request by beridel**
Hello,

I've implemented a Neumaier sum to do the cumulative sum (https://en.wikipedia.org/wiki/Kahan_summation_algorithm). That way the normalization of the data in the correlation coefficient function only needs to do a subtraction to get the sum of the squares of the data.

I've also fixed an outstanding problem with the CPU version: the cumulative sum of squares was being calculated out of the matched_filter call. Now everything happens at once. This means there's no difference between the Matlab and python calls now.

There seems to be a strange bug with the python version when you use a step that's more than 1... I'm not sure yet where that is coming from; let me know if you have any ideas.

Otherwise, let me know if you can break it!

NB: Of course this only affects the CPU version of FMF.

**Edited to add the link.**

### CODE REVIEW

**Comment by ebeauce** [comment]
Hi William,

I think we've already investigated how the Kahan summation algorithm could help us to get an accurate cumulative sum using single precision float numbers. 

Continuous data with high amplitudes are still going to produce cumulative sum with numbers that cannot be well represented with only 4 bytes, and the subtraction to get the denominator is still going to be affected by this lack of accuracy. For example:

     single_2 = np.float32(1.e9 + 1234.4573745678) 
     single_1 = np.float32(1.e9 + 1234.4573745678 - 1000) #single_2 - single_1 should trivially yields 1000
     print(single_2 - single_1) #the result is 960 
    double_2 = np.float64(1.e9 + 1234.4573745678)
    double_1 = np.float64(1.e9 + 1234.4573745678 - 1000)
    print(double_2 - double_1) # the result is 1000 

discrepancy

This first figure shows the CCs computed with the standard fmf (blue) and the neumaier fmf (orange)
 at the end of the day. Even though the beginning of the day is similar for both codes, the end of the day shows a large discrepancy between the two.

cum_error

Another example showing that the cumulative sum on single float numbers is very likely to break down for high amplitudes / long time series is coming from the kaikoura dataset (in the tests directory). I plot the cumulative different (CC_standard - CC_neumaier)**2 and we can see that the error blows out at some point.

To summarize, no matter what is the accuracy of the summation algorithm, storing large numbers as float 32 is always going to be associated with loss of precision. In the current implementation, the cumulative sum is computed with float 64 numbers and the subtraction is performed for each sliding window. Subtraction of the cumulative sum yields small numbers that can in turn be converting to float32 without important loss of precision.

I would be glad to be wrong somewhere, but unfortunately I really think this is a limitation due to single precision that cannot be overcome.

**Comment by beridel** [comment]
I forgot about that damn problem. I was just too excited to make the Neumaier summation work.

So I've satisfied your passion for doubles with the latest commit. Let me know what you think and if it passes the test.

And unless I'm mistaken, the current version doesn't do a subtraction of the cumulative sum to compute the data sum of squares; it just computes the sum of squares for every sample(step). What I'd be interested to see if this new method goes any faster than the current implementation. It most likely depends on how many cores you can throw at the current implementation. Maybe a comparison between a single core for both implementations?

**Comment by ebeauce** [comment]
Ok, using doubles seems to solve the issue ! 

These are the same figures as in my previous message, but with the new code:

gpu_vs_neumaier

The CC time series matches well with the CCs computed by the GPU version.

cum_error

The cumulative error increases roughly linearly, as we would expect from the summation of small errors.

Also, I compared the run times between the Neumaier implementation and the current CPU implementation, and the Neumaier version was way faster with my data. I tried to use the test function to get an idea of how the speedup scales when increasing the number of templates/stations/components, and realized that a few variables should be converted to int (like n_samples_data and n_samples_template). 

Parallelizing the station loop in cumsum_square_data with a simple 
#pragma omp parallel for private(s, c, d, station_offset, component_offset)
didn't improve a lot the run time, but worked well (have to look what is the advantage when using many stations / components). In the same function, the line
if (weights[component_offset] == 0) continue;
should be removed, as weights is a matrix that is defined for each template, so checking this condition is wrong since the cumulative sum is used by every template.

File: fast_matched_filter/src/matched_filter.c [hunk 1]
```diff
@@ -112,29 +120,55 @@ float corrc(float *templates, float sum_square_template,
...
+                        double *csum_square_data) {
+    int s, c;
+    int station_offset, component_offset, d;
+
+    for (s = 0; s < n_stations; s++) {
+        station_offset = s * n_components;
+
+        for (c = 0; c < n_components; c++) {
+            component_offset = station_offset + c;
+            if (weights[component_offset] == 0) continue;
```
Comment (ebeauce) [inline fast_matched_filter/src/matched_filter.c]: It's not correct to check this condition, because weights is not the same for different templates.

File: fast_matched_filter/src/matched_filter.c [hunk 2]
```diff
@@ -112,29 +120,55 @@ float corrc(float *templates, float sum_square_template,
...
-        // sliding cumulative sum
-#pragma omp parallel for private(ch, i, n) shared(data_sq_ch, csum_ch)
-        for (n = 0; n < n_samples_data-n_samples_template; n++){
-            for (i = 0; i < n_samples_template; i++) csum_ch[n] += data_sq_ch[n+i];
+void cumsum_square_data(float *data, int n_samples_data, float *weights,
+                        int n_stations, int n_components,
+                        double *csum_square_data) {
+    int s, c;
+    int station_offset, component_offset, d;
+
```
Comment (ebeauce) [inline fast_matched_filter/src/matched_filter.c]: #pragma omp parallel for private(s, c, d, station_offset, component_offset)
is a cheap way of optimizing this function

**Review by ebeauce — CHANGES_REQUESTED** [review]
I added two comments about the "if" condition with the weights, and the addition of a pragma parallel statement.

I don't know how to include comments on lines 192 and 199 of fast_matched_filter.py (in the test function), but these variables have to be converted explicitly to integers.

**Comment by calum-chamberlain** [comment]
Hey chaps

Brief note from me without actually looking at the code... Eric mentioned using ints for number of data samples, in my experience these need to be long ints to avoid overflow for large numbers.

File: fast_matched_filter/src/matched_filter.c (hunk 1 above)
Comment (beridel) [reply fast_matched_filter/src/matched_filter.c]: Good point.

File: fast_matched_filter/src/matched_filter.c (hunk 2 above)
Comment (beridel) [reply fast_matched_filter/src/matched_filter.c]: In an upcoming commit, I abstracted stations and components to channels and looped over that. More parallel = better!

**Comment by beridel** [comment]
For the explicit integer casting: this is handled in the matched_filter function itself. So I don't think we need to take care of it in the test function.

I've implemented the pragma line as suggested, though I parallelized over channels and not stations or components.

On my 10-core iMac, it saves about 7 seconds to compute the data sum of squares. I'm guessing this time difference will be even bigger on machines with less cores?

Calum: concerning the long ints, when have you run into this problem? What giant number is big enough to explode a 32-bit int?

In any case, I think it's good to deploy. Let me know what you think.

**EDIT**: with a step greater than 1, there is little chance of generate templates at a multiple of the step. I made it so the "random" template times align to the step.

**Comment by calum-chamberlain** [comment]
TBH, I can't remember - you would have to be getting up beyond the 2x10^10 mark to run into it as an issue - might have only been an issue for us when we were doing strange things with the fft's in EQcorrscan (if it's just the number of samples, then the largest int is something like 248 days at 100Hz: Probably plenty!).

File: fast_matched_filter/fast_matched_filter.py [hunk 3]
```diff
@@ -12,7 +12,7 @@
 import ctypes as ct
 import datetime as dt
 import os
-
+from IPython.core.debugger import Tracer; debug_here = Tracer()
```
Comment (calum-chamberlain) [inline fast_matched_filter/fast_matched_filter.py]: If this is going to be left in you should add ipython to your requirements [here](https://github.com/beridel/fast_matched_filter/blob/master/setup.py#L69)

File: fast_matched_filter/src/matched_filter.c [hunk 4]
```diff
@@ -58,19 +64,21 @@ void matched_filter(float *templates, float *sum_square_templates, int *moveouts
...
+                                                        csum_square_data + i + 1,
                                                         weights_t,
                                                         n_samples_template,
                                                         n_samples_data,
                                                         n_stations,
                                                         n_components);
         }
     }
+
+    free(csum_square_data);
```
Comment (calum-chamberlain) [inline fast_matched_filter/src/matched_filter.c:76]: Do you need to free `template_t`, `sum_square_templates_t`, `weights_t` and `moveouts_t` as well?

File: fast_matched_filter/src/matched_filter.c [hunk 5]
```diff
@@ -112,29 +120,50 @@ float corrc(float *templates, float sum_square_template,
     for (i = 0; i < n_samples_template; i++){
         numerator += templates[i] * data[i];
     }
-    denominator = sum_square_template * csum_square_data[0];
+    denominator = sum_square_template * (float)(csum_square_data[n_samples_template] - csum_square_data[-1]);
```
Comment (calum-chamberlain) [inline fast_matched_filter/src/matched_filter.c:123]: I think I probably would have done this as double precision math (e.g. have `sum_square_template` as a double, and cast to float after the multiplication - but if this isn't causing issues, its probably okay.

File: fast_matched_filter/src/matched_filter.c [hunk 6]
```diff
@@ -112,29 +120,50 @@ float corrc(float *templates, float sum_square_template,
...
-            for (i = 0; i < n_samples_template; i++) csum_ch[n] += data_sq_ch[n+i];
+void cumsum_square_data(float *data, int n_samples_data, float *weights,
+                        int n_stations, int n_components,
+                        double *csum_square_data) {
+    int ch, d;
+
+    // loop over channels
+#pragma omp parallel for private(ch, d)
+    for (ch = 0; ch < n_stations * n_components; ch++) {
+        d = ch * n_samples_data;
```
Comment (calum-chamberlain) [inline fast_matched_filter/src/matched_filter.c:139]: This may be where large integers become an issue - say if you have a lot of channels (250... not sure anyone has done that large a correlation run?) and a day of data, you would run into overflow. The cost of using a 64Bit int here is probably pretty low and may be safer? Up to you though, I think the overflow case is pretty unlikely.

**Review by calum-chamberlain — COMMENTED** [review]
Just a few comments that you can probably take or leave as you see fit.

**Comment by ebeauce** [comment]
It might be good to systematically change int to long int.

Otherwise, we do have to use integer variables when we define the dimensions of a numpy array. On my computer, the test function doesn't work because n_samples_data and n_samples_template are floats.

File: fast_matched_filter/fast_matched_filter.py (hunk 3 above)
Comment (beridel) [reply fast_matched_filter/fast_matched_filter.py]: Good catch!

File: fast_matched_filter/src/matched_filter.c (hunk 4 above)
Comment (beridel) [reply fast_matched_filter/src/matched_filter.c:76]: No, all of those are just pointers to some other array. I never need to allocate any memory to them.

Comment (calum-chamberlain) [reply fast_matched_filter/src/matched_filter.c:76]: Ah, jolly good! Apologies for the distraction!

File: fast_matched_filter/src/matched_filter.c (hunk 5 above)
Comment (beridel) [reply fast_matched_filter/src/matched_filter.c:123]: I just played around with all sorts of combinations, even going so far as to make the sum of squares of the template a double as well, but it doesn't seem to affect the accuracy at all....

File: fast_matched_filter/fast_matched_filter.py (hunk 3 above)
Comment (beridel) [reply fast_matched_filter/fast_matched_filter.py]: Good catch...

File: fast_matched_filter/src/matched_filter.c (hunk 6 above)
Comment (beridel) [reply fast_matched_filter/src/matched_filter.c:139]: Let's get this merged and then we can do another update for that. This is already a big jump forward with the Neumaier sum and the new implementation of the cumulative sum in the C code (and not called by Matlab or python).

**Comment by beridel** [comment]
Let me know if you guys are good to merge.

**Comment by ebeauce** [comment]
I am good to merge, except that I still think that in the test function we should define n_samples_data and n_samples_template such that they are forced to be integer variables (they currently are floats):

n_samples_data = np.int32( ------- )
n_samples_template = np.int32( ------ )

Otherwise I am afraid that this prevents the function to work (as with my current python environment).

**Review by ebeauce — APPROVED** [review]

**Comment by beridel** [comment]
Merged.

Re Eric's comment on forcing integers: lines 98/99 take care of that already, no?

**Comment by ebeauce** [comment]
I was thinking of the test function, i.e. lines 194 and 201.

I'll create a pull request to correct that :)

**Comment by beridel** [comment]
But those variables don't even make it into the matched_filter function. Numpy will already check for them to be integers (like if the user gives some silly decimal duration) when the call to np.zeros is made.
